### ISSUE TITLE
[Meta] Validate plugins compatibility with segment replication

### ISSUE THREAD

**Issue by dreamer-89** [opensearch-project/OpenSearch#8211]
### Summary
With 2.9.0 release, there are lot of enhancements going in for segment replication[1][2] feature (went GA in 2.7.0), we need to ensure different plugins are compatible with current state of this feature. Previously, we ran [tests](https://github.com/opensearch-project/OpenSearch/issues/6602) on plugin repos to verify this compatibility but want plugin owners to be aware of these changes so that required updates (if any) can be made. With `2.10.0` release, [remote store feature](https://github.com/opensearch-project/OpenSearch/issues/2992) is going GA which internally uses SEGMENT replication strategy only i.e. it enforces all indices to use `SEGMENT` replication strategy. So, it is important to validate plugins are compatible with segment replication feature. 

#### 1. Refresh policy behavior
1. RefreshPolicy.IMMEDIATE will only refresh primary shards but not replica shards immediately. Instead post refresh, primary will start  a round of segment replication to update the replica shard copies leading to eventual consistency.
2. RefreshPolicy.WAIT_UNTIL ensures the indexing operation is searchable in your cluster i.e. RAW (Read after write guarantee). With segment replication, this guarantee is not promised due to delay in replica shared updates from asynchronous background refreshes.

####  2. Refresh lag on replicas
With segment replication, there is inherent delay in documents to be searchable on replica shard copies. This is due to the fact that replica shard copies over data (segment) files from primary. Thus, compared to document replication, there will be on average increase in amount of time the replica shards are consistent with primaries. 

#### 3. System/hidden indices support
With https://github.com/opensearch-project/OpenSearch/pull/8200, system and hidden indices are now supported with `SEGMENT` replication strategy. We need to ensure there are no bottlenecks which prevents system/hidden indices with segment replication.

### Next steps
With segment replication strong reads are not guaranteed. Thus, if the plugin needs strong reads guarantees specially as alternative to change in behavior of refresh policy and lag on replicas (point 1 and 2 above), we need to update search requests to target primary shard only. With https://github.com/opensearch-project/OpenSearch/pull/7375, core now supports primary shards only based search. Please follow [documentation](https://opensearch.org/docs/latest/api-reference/search/#the-preference-query-parameter) for examples and details 

### Open questions
In case of any questions or issues, please post your question on [core](https://github.com/opensearch-project/OpenSearch/issues/8211)
 
### Reference
[1] [Design](https://github.com/opensearch-project/OpenSearch/issues/2229)
[2] [Documentation](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/segment-replication/index/)

Opensearch Plugins
- [x] https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4444 
   - [x] https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4522
- [ ] https://github.com/opensearch-project/alerting/issues/974
- [x] https://github.com/opensearch-project/anomaly-detection/issues/936
  - [x] https://github.com/opensearch-project/anomaly-detection/issues/989
- [x] https://github.com/opensearch-project/asynchronous-search/issues/303
- [x] https://github.com/opensearch-project/common-utils/issues/469
- [x] https://github.com/opensearch-project/reporting/issues/719
- [x] https://github.com/opensearch-project/index-management/issues/833
- [x] https://github.com/opensearch-project/job-scheduler/issues/407
  - [x] https://github.com/opensearch-project/job-scheduler/issues/471
- [x] https://github.com/opensearch-project/k-NN/issues/948
- [x] https://github.com/opensearch-project/performance-analyzer/issues/491
- [x] https://github.com/opensearch-project/performance-analyzer-rca/issues/433
- [x] https://github.com/opensearch-project/security/issues/2916
- [x] https://github.com/opensearch-project/sql/issues/1801
- [x] https://github.com/opensearch-project/observability/issues/1562
- [x] https://github.com/opensearch-project/cross-cluster-replication/issues/1029
- [x] https://github.com/opensearch-project/ml-commons/issues/1023
- [x] https://github.com/opensearch-project/geospatial/issues/346
- [x] https://github.com/opensearch-project/notifications/issues/697
- [x] https://github.com/opensearch-project/neural-search/issues/206
- [x] https://github.com/opensearch-project/security-analytics/issues/473
- [x] https://github.com/opensearch-project/opensearch-oci-object-storage/issues/52

OpenSearch-dashboard plugins
- [x] https://github.com/opensearch-project/dashboards-query-workbench/issues/88
- [x] https://github.com/opensearch-project/alerting-dashboards-plugin/issues/587
- [x] https://github.com/opensearch-project/dashboards-notifications/issues/64
- [x] https://github.com/opensearch-project/security-dashboards-plugin/issues/1494
- [x] https://github.com/opensearch-project/index-management-dashboards-plugin/issues/798
- [x] https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/issues/520
- [x] https://github.com/opensearch-project/dashboards-reporting/issues/132
- [x] https://github.com/opensearch-project/dashboards-observability/issues/586
- [x] https://github.com/opensearch-project/dashboards-visualizations/issues/196
- [x] https://github.com/opensearch-project/dashboards-search-relevance/issues/228
- [x] https://github.com/opensearch-project/dashboards-maps/issues/430
- [x] https://github.com/opensearch-project/security-analytics-dashboards-plugin/issues/639
- [x] https://github.com/opensearch-project/ml-commons-dashboards/issues/226

**Comment by mch2** [comment]
Related, remote store compat with system indices is broken until https://github.com/opensearch-project/OpenSearch/issues/8193 is resolved.

**Comment by dreamer-89** [comment]
@opensearch-project/admin : Request to add `v2.9.0` labels to above plugins issues.

**Comment by minalsha** [comment]
@dreamer-89 , @Rishikesh1159  can you please share a sample test that we can refer to which we need to add to create system indices and verify that if system/hidden indices in a given plugin work correctly with segment replication?

**Comment by dreamer-89** [comment]
> @dreamer-89 , @Rishikesh1159 can you please share a sample test that we can refer to which we need to add to create system indices and verify that if system/hidden indices in a given plugin work correctly with segment replication?

Thanks you @minalsha for your question. The [documentation](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/segment-replication/index/) (also mentioned in issue above) describes different ways you can create segment replication enabled indices. Please have a look and let me know if there are more questions.

**Comment by dbwiddis** [comment]
> To enable segment replication at index level, we can pass in `replication.type` as `SEGMENT`. Please note, it overrides the cluster level setting.
> 
> ```
> curl -X PUT "$host/{index_name}" -H 'Content-Type: application/json' -d '{
>   "settings": {
>     "index": {
>       "number_of_shards": 1,
>       "number_of_replicas": 1, 
>       "replication.type": "SEGMENT"
>     }
>   }
> }'
> ```

Hey @dreamer-89 I'm trying to use this programmatically when creating a system index to make sure `ReplicationType` is always `DOCUMENT`, and want to make sure I'm reading this correctly.

The current code uses `CreateIndexRequest`:
```java
CreateIndexRequest request = new CreateIndexRequest(LOCK_INDEX_NAME) ...
```
So in my case if I create a settings object to pass a setting "replication.type" (which is probably a constant somewhere) of "DOCUMENT" (which is an enum) this now becomes:
```java
CreateIndexRequest request = new CreateIndexRequest(LOCK_INDEX_NAME, replicationSettings) ...
```
this would always have that index set document even if the user chose SEGMENT for their cluster, correct?

**Comment by dreamer-89** [comment]
> > To enable segment replication at index level, we can pass in `replication.type` as `SEGMENT`. Please note, it overrides the cluster level setting.
> > ```
> > curl -X PUT "$host/{index_name}" -H 'Content-Type: application/json' -d '{
> >   "settings": {
> >     "index": {
> >       "number_of_shards": 1,
> >       "number_of_replicas": 1, 
> >       "replication.type": "SEGMENT"
> >     }
> >   }
> > }'
> > ```
> 
> Hey @dreamer-89 I'm trying to use this programmatically when creating a system index to make sure `ReplicationType` is always `DOCUMENT`, and want to make sure I'm reading this correctly.
> 
> The current code uses `CreateIndexRequest`:
> 
> ```java
> CreateIndexRequest request = new CreateIndexRequest(LOCK_INDEX_NAME) ...
> ```
> 
> So in my case if I create a settings object to pass a setting "replication.type" (which is probably a constant somewhere) of "DOCUMENT" (which is an enum) this now becomes:
> 
> ```java
> CreateIndexRequest request = new CreateIndexRequest(LOCK_INDEX_NAME, replicationSettings) ...
> ```
> 
> this would always have that index set document even if the user chose SEGMENT for their cluster, correct?

Thanks @dbwiddis for your comment. Yes index level `replication.type` setting overwrites the cluster level setting. Though the ask here is to verify plugin compatibility with segment replication but not to overcome it by using `DOCUMENT` as replication strategy. Going forward `SEGMENT` will be the only replication strategy supported for certain configurations (e.g. Remote store at cluster level). Thus, there is no way other than to validate indices actually created with `SEGMENT` replication.

**Comment by scrawfor99** [comment]
To echo @dbwiddis, I am concerned with the effect that these changes will have on other sensitive system index use cases. With the current security model, we make use of a system index to store important user information such as internal role mapping. With https://github.com/opensearch-project/security/pull/2903, we appear to handle the consistency issue which arises with the replication. However, we still don't account for the performance impact of copying  the system index repeatedly. We already run into performance concerns due to reloading the nodes every time there is a configuration change related to Security: https://github.com/opensearch-project/security/issues/1842. With this change, I am concerned we would be doubling the cost as we now would need to update the duplicated system index frequently. 

Perhaps I am conflating the processes at hand here, but I do think we should aim for verifiable performance metrics before making such a significant change.

**Comment by seankao-az** [comment]
I have a question on this:
> With segment replication strong reads are not guaranteed. Thus, if the plugin needs strong reads guarantees specially as alternative to change in behavior of refresh policy and lag on replicas (point 1 and 2 above), we need to update search requests to target primary shard only.

In case of primary shard failure, does a replica get all the update before it gets promoted to primary? Otherwise strong read isn't guaranteed even if we read only from primary.

**Comment by anasalkouz** [comment]
@seankao-az 
Recovery scenarios are covered on SegRep and you should't face any data loss. We guarantee  data durability by storing all operation in translog, which means incase of primary shard failure, replica can replay all remaining operations before being promoted as primary

**Comment by mch2** [comment]
Today with document replication only way to get strong reads is with GET request by ID (which reads from translog independent of refresh cycle) or use the RefreshPolicy.WAIT_UNTIL during ingest followed by issuing a search request.

Segment Replication does not currently support either of these mechanisms. With WAIT_UNTIL the write would be ack'd only after the primary has made the docs searchable, the replicas would still be in the process of syncing.  To achieve strong read after this request is made the search would also need to prefer primary to achieve a strong read.

There is no reason we can't support get by id. This is a miss - I made an [issue](https://github.com/opensearch-project/OpenSearch/issues/8536) here to support this.

> In case of primary shard failure, does a replica get all the update before it gets promoted to primary? Otherwise strong read isn't guaranteed even if we read only from primary.

Yes, in failover case the new primary catches up before serving reads or new writes.

**Comment by dreamer-89** [comment]
> > Though the ask here is to verify plugin compatibility with segment replication but not to overcome it by using DOCUMENT as replication strategy.
> 
> @dreamer-89 Understood. To be clear:
> 
> 1. I'm tracking the ask to verify compatibility.
> 2. I'm asking about the best way to avoid the known performance regression that this is introducing.  There has been no testing to measure how much the performance will degrade, just a statement "System indices are generally small and should copy out quickly, so I wouldn't expect this to be of major concern."
> 
> In Job Scheduler, a system index is used for locking in highly parallel, performance-sensitive use cases. We absolutely need a way for users in these use cases to change this. And while the "change the index setting" works for other indices, changing settings on system indices has more safeguards.
> 
> I take your comment above "`SEGMENT` will be the only replication strategy supported for certain configurations" to mean that my proposed hard-coded forcing of this setting is wrong.
> 
> What, then, is the correct action for a performance-sensitive application that does not want WAIT_UNTIL on their system index to take here?


Hi @dbwiddis,
Thanks for sharing your use-case and apologies for delay on my end. 

Yes, if [primary shard based searching](https://opensearch.org/docs/latest/api-reference/search/#the-preference-query-parameter) does not work for you, then your plan as mentioned in job-scheduler [issue](https://github.com/opensearch-project/job-scheduler/issues/407#issuecomment-1627834721) sounds right to me.

**Comment by dreamer-89** [comment]
Closing this issue as all plugins except alerting have done necessary changes for compatibility with SEGMENT replication feature. For alerting, created a separate issue on core to fix https://github.com/opensearch-project/OpenSearch/issues/9669 and created documentation issue to mark this as limitation in https://github.com/opensearch-project/documentation-website/issues/4967

CC @anasalkouz @mch2
